Add pkgbuild option to build local package

ckotte opened the pull request:
Sometimes AUR packages are broken and not fixed. With the pkgbuild option, a local directory with the PKGBUILD and build files can be specified to build the package.

kewlfft commented:
I think it is a nice feature, thanks for your contribution. I have some comments:

- The option name `pkgbuild` is used by _pikaur_, but wouldn't `local` be more self explanatory?

- I am fine with your _makepkg_ implementation of it 

- I don't want to be too specific about _pikaur_ and have a specific function for it. Other helpers may have some ways to achieve the same now or in the future (see https://github.com/Jguer/yay/issues/694 and https://github.com/Jguer/yay/issues/647) and I would prefer to keep the tool as generic and scalable as possible.

ckotte commented:
I could rename it to ```local_pkgbuild```.

I use pikaur in my playbook and I have a specific use case for local pkgbuilds. I just added the functionality for makepkg after I implemented it for pikaur. 

Other helpers cannot build local pkgbuilds yet. You could change the function in the future and keep the same options for pikaur and add the commands for the other helpers. 

Or you wait until every helper implements it and then you can add it to your module. However, this probably won't happen. pacaur is also discontinued and the issue for yay is open since 2 years.

ckotte commented:
Are you happy if I change the new pikaur function to a generic function where someone could add the functionality for other helpers in the future??

kewlfft commented:
yes adding or for removing a helper should require minimum maintenance, it often happens. We have to think if _pikaur_ is discontinued, it should be a question of adding or removing an element in a list.

ckotte commented:
I changed the code. Now there's a list for a local package build and a generic function that assumes the last argument is the PKGBUILD file. Both are only used if ```local_pkgbuild``` is specified. 

This will work in theory with another helper that just need to be added to the new list. In theory because it will only work if the helper requires the PKGBUILD file as the last argument. It doesn't work anymore if the helper requires the directory or if it needs to execute the command inside the directory, etc. pp.

The code is now more bloated. Tell me what you think. I will rebase the PR later.

kewlfft commented:
it seems good thanks for the changes

ckotte commented:
ok. rebase done.

kewlfft commented:
can't `extra_args` be used to add the remove dependencies options to makepkg ?

kewlfft commented:
for behavior that varies depending on the helper let's not try to change or harmonize in the background, this would add complexity. `--skip-pgp-check` was implemented before `extra_args` which was only added last June. If `extra_args` does the job, we could maybe remove the skip pgp check (breaking change).

ckotte commented:
I removed the last commit. Sorry. I don't have time for this.

ckotte commented:
I would not remove this option because there's also a description that explicitly says why it's useful in the first place. This helps users if they have exactly this issue/use case of installing only one package (AUR helper) that needs GnuPG configured.

soloturn commented:
i'd love if the option "yay -U" learns to take either a package file and installs it, or PKGBUILD and builds it beforehand, instead of having "local_pkgbuild". otherwise, thanks a lot for this contribution!

kewlfft commented:
@soloturn what you describe and would like _yay_ to support is it what _pikaur_ does already with the `pkgbuild` option?

kewlfft commented on library/aur.py:
```diff
@@ -9,6 +9,7 @@
 import tarfile
 import os
 import os.path
+from distutils.dir_util import copy_tree
```
We need a good reason to use this instead of shutil's copytree()

ckotte replied:
What's the exact problem of using copy_tree from distutils instead of copytree from shutil?

kewlfft replied:
This is not the standard package for file operations, not sure how well this function will be maintained and why we want to rely on a distribution package

kewlfft commented on library/aur.py:
```diff
@@ -102,6 +111,11 @@
 
+use_cmd_local_pkgbuild = {
+    'pikaur': ['pikaur', '-P', '--noconfirm', '--noedit', '--needed', '--install'],
+    'makepkg': ['makepkg', '--syncdeps', '--rmdeps', '--install', '--noconfirm', '--needed']
```
I would remove `--rmdeps` here, it can be added with extra_args if needed
```suggestion
    'makepkg': ['makepkg', '--syncdeps', '--install', '--noconfirm', '--needed']
```

kewlfft commented on library/aur.py:
```diff
@@ -301,6 +346,15 @@ def make_module():
     if use != 'makepkg' and (params['skip_pgp_check'] or params['ignore_arch']):
         module.fail_json(msg="This option is only available with 'makepkg'.")
 
+    if not (use == 'makepkg' or use == 'pikaur') and params['local_pkgbuild']:
```
test if use is in use_cmd_local_pkgbuild instead

kewlfft commented on library/aur.py:
```diff
@@ -145,39 +159,61 @@ def check_packages(module, packages):
+        f = open_url('https://aur.archlinux.org/{}'.format(result['URLPath']))
+    with tempfile.TemporaryDirectory() as tmpdir:
+        if local_pkgbuild:
+            copy_tree(local_pkgbuild, tmpdir)
```
why not the regular `copytree()` ?

ckotte replied:
I get ```FileExistsError: [Errno 17] File exists: '/tmp/tmp_bkpq7iv'``` with ```copytree```. There are no issues with ```copy_tree```..

kewlfft replied:
The destination directory must not already exist. it behaves differently.

ckotte replied:
ok. this works.

ckotte commented:
I implemented all suggestions. Please check.
